qiskit/dagcircuit/_dagcircuit.py

```diff
@@ -997,181 +1012,291 @@ def nodes(self):
         return nodes
 
     def get_gate_nodes(self, data=False):
         """Deprecated. Use gate_nodes()."""
-        warnings.warn('The method get_gate_nodes() is being replaced by gate_nodes()',
+        warnings.warn('The method get_gate_nodes() is being replaced by gate_nodes().'
+                      'Returning a list of node_ids/(node_id, data) tuples is also '
+                      'deprecated, gate_nodes() returns a list of DAGNodes ',
                       DeprecationWarning, 2)
-        return self.gate_nodes(data)
+        if data:
+            warnings.warn('The parameter data is deprecated, '
+                          'get_gate_nodes() now returns DAGNodes '
+                          'which always contain the data',
+                          DeprecationWarning, 2)
 
-    def gate_nodes(self, data=False):
-        """Get the list of gate nodes in the dag.
+        nodes = []
+        for node in self.op_nodes():
+            if isinstance(node['op'], Gate):
+                nodes.append((node._node_id, node))
+        if not data:
+            nodes = [n[0] for n in nodes]
+        return nodes
 
-        Args:
-            data (bool): Default: False. If True, return a list of tuple
-                (node_id, node_data). If False, return a list of int (node_id)
+    def gate_nodes(self):
+        """Get the list of gate nodes in the dag.
 
         Returns:
             list: the list of node ids that represent gates.
         """
         nodes = []
-        for node_id, node_data in self.op_nodes(data=True):
-            if isinstance(node_data['op'], Gate):
-                nodes.append((node_id, node_data))
-        if not data:
-            nodes = [n[0] for n in nodes]
+        for node in self.op_nodes():
+            if isinstance(node.op, Gate):
+                nodes.append(node)
         return nodes
 
     def get_named_nodes(self, *names):
         """Deprecated. Use named_nodes()."""
         warnings.warn('The method get_named_nodes() is being replaced by named_nodes()',
+                      'Returning a list of node_ids is also deprecated, named_nodes() '
+                      'returns a list of DAGNodes ',
                       DeprecationWarning, 2)
-        return self.named_nodes(*names)
+
+        named_nodes = []
+        for node in self.multi_graph.nodes():
+            if node['type'] == 'op' and node['op'].name in names:
+                named_nodes.append(node._node_id)
+        return named_nodes
 
     def named_nodes(self, *names):
         """Get the set of "op" nodes with the given name."""
         named_nodes = []
-        for node_id, node_data in self.multi_graph.nodes(data=True):
-            if node_data['type'] == 'op' and node_data['op'].name in names:
-                named_nodes.append(node_id)
+        for node in self.multi_graph.nodes():
+            if node.type == 'op' and node.op.name in names:
+                named_nodes.append(node)
         return named_nodes
```

Comment:
can this be an iterator?

Reply:
I tried it out but passes `barrier_before_final_measurements` and `cx_direction` modify the list (deleting elements/adding them) during the pass and generators don't support this. It might be possible but I think it should be a separate issue.